meilisearch-http/src/routes/search.rs

```diff
@@ -1,17 +1,128 @@
-use std::collections::{BTreeSet, HashSet};
+use std::any::{Any, TypeId};
+use std::collections::{BTreeSet, HashMap, HashSet};
+use std::marker::PhantomData;
+use std::ops::Deref;
 
-use actix_web::{get, post, web, HttpResponse};
+use actix_web::{web, FromRequest, HttpResponse};
+use futures::future::{err, ok, Ready};
 use serde::Deserialize;
 use serde_json::Value;
 
-use crate::error::ResponseError;
-use crate::helpers::Authentication;
-use crate::index::{SearchQuery, default_crop_length, DEFAULT_SEARCH_LIMIT};
+use crate::error::{AuthenticationError, ResponseError};
+use crate::index::{default_crop_length, SearchQuery, DEFAULT_SEARCH_LIMIT};
 use crate::routes::IndexParam;
 use crate::Data;
 
+struct Public;
+
+impl Policy for Public {
+    fn authenticate(&self, _token: &[u8]) -> bool {
+        true
+    }
+}
+
+struct GuardedData<T, D> {
+    data: D,
+    _marker: PhantomData<T>,
+}
+
+trait Policy {
+    fn authenticate(&self, token: &[u8]) -> bool;
+}
+
+struct Policies {
+    inner: HashMap<TypeId, Box<dyn Any>>,
```

Comment:
@irevoire maybe we could use [phf](https://docs.rs/phf/0.9.0/phf/)

Reply:
Yes, we could. It would be easier to use. We could also use a `lazy_static` if we want to keep using a regular `HashMap` :eyes: 
Do you happen to know if we can create `HashMap` in `const fn` yet? It would be a cool third alternative also.

But honestly, do what you prefer. I don't have any problem with your solution :grin:

Reply:
I decided to leave it like that for the moment.